### PR TITLE
feat: refactor redispool

### PR DESCRIPTION
**Pull request by daheige**
**Please provide issue(s) of this PR:**
refactor #
refactor common/redispool
for https://github.com/polarismesh/polaris/issues/404
for https://github.com/polarismesh/polaris/pull/403
for https://github.com/polarismesh/polaris/pull/398

### CODE REVIEW

**Comment by andrewshan** [comment]
@[magederek](https://github.com/magederek), it may be conflicted with your pr, pls check it #405

**Comment by magederek** [comment]
Good style for options! As my opinion, not all the options from goredis should be revealed such as tls.Config.
Please also update the test files, yaml configs and helm charts for a full configuration change. 

> @[magederek](https://github.com/magederek), it may be conflicted with your pr, pls check it #405

#403 and #405 are given some moderate modification for #398 and #404. Refactor of redispool options's code should be an independent issue to discuss and enhance, which can be merge later.

File: common/redispool/redis_pool.go [hunk 1]
```diff
@@ -93,31 +93,96 @@ type Resp struct {
...
+	// TLS Config to use. When set TLS will be negotiated.
+	TlsConfig *tls.Config
 }
 
 // DefaultConfig redis pool configuration with default values
 func DefaultConfig() *Config {
 	return &Config{
 		MaxIdle:        200,
-		IdleTimeout:    commontime.Duration(120 * time.Second),
-		ConnectTimeout: commontime.Duration(300 * time.Millisecond),
```
Comment (andrewshan) [inline common/redispool/redis_pool.go:114]: 默认连接时长和消息超时为什么要改成5s？

File: common/redispool/config_option.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+	commontime "github.com/polarismesh/polaris-server/common/time"
+)
+
+// Option functional options for Config
+type Option func(c *Config)
+
+// WithConfig set new config for NewPool,keep old code compatibility
+func WithConfig(newConfig *Config) Option {
+	return func(c *Config) {
+		c = newConfig
```
Comment (chuntaojun) [inline common/redispool/config_option.go:16]: 这个逻辑，应该是问题的？并不能真正影响c的实际内容吧

File: common/redispool/redis_pool.go [hunk 3]
```diff
@@ -93,31 +93,96 @@ type Resp struct {
...
+	// to a Redis 6.0 instance, or greater, that is using the Redis ACL system.
+	Username string `json:"username"`
+
+	// WithTLS whether open TLSConfig
+	// if WithTLS is true, you should call WithEnableWithTLS,and then TLSConfig is not should be nil
+	// In this case you should call WithTLSConfig func to set tlsConfig
+	WithTLS bool `json:"withTLS"`
+
+	// TLS Config to use. When set TLS will be negotiated.
+	TlsConfig *tls.Config
```
Comment (magederek) [inline common/redispool/redis_pool.go:166]: both tlsConfig itself and fields in tls.Config have not `json` tag for unmarshalling from JSON config.

**Comment by daheige** [comment]
There is no problem with the design here. You can understand it by looking carefully at the design of with options


------------------ Original ------------------
From: liaochuntao ***@***.***&gt;
Date: Sun,Jun 5,2022 10:37 PM
To: polarismesh/polaris ***@***.***&gt;
Cc: heige ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [polarismesh/polaris] feat: refactor redispool (PR #408)





 
@chuntaojun commented on this pull request.
 
 
In common/redispool/config_option.go:
 &gt; +package redispool + +import ( +"crypto/tls" +"time" + +commontime "github.com/polarismesh/polaris-server/common/time" +) + +// Option functional options for Config +type Option func(c *Config) + +// WithConfig set new config for NewPool,keep old code compatibility +func WithConfig(newConfig *Config) Option { +return func(c *Config) { +c = newConfig  
这个逻辑，应该是问题的？并不能真正影响c的实际内容吧

**Comment by daheige** [comment]
This is a private attribute. It is provided to the upstream of the business for configuration. It does not need to be exposed here. Moreover, each TLS configuration is optional for users, so it is designed in this way. See with options.


------------------ Original ------------------
From: Derek.Chen ***@***.***&gt;
Date: Sun,Jun 5,2022 10:39 PM
To: polarismesh/polaris ***@***.***&gt;
Cc: heige ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [polarismesh/polaris] feat: refactor redispool (PR #408)





 
@magederek commented on this pull request.
 
 
In common/redispool/redis_pool.go:
 &gt; +// Connection age at which client retires (closes) the connection. +// Default is to not close aged connections. +MaxConnAge commontime.Duration `json:"maxConnAge"` + +// Use the specified Username to authenticate the current connection +// with one of the connections defined in the ACL list when connecting +// to a Redis 6.0 instance, or greater, that is using the Redis ACL system. +Username string `json:"username"` + +// WithTLS whether open TLSConfig +// if WithTLS is true, you should call WithEnableWithTLS,and then TLSConfig is not should be nil +// In this case you should call WithTLSConfig func to set tlsConfig +WithTLS bool `json:"withTLS"` + +// TLS Config to use. When set TLS will be negotiated. +TlsConfig *tls.Config  
both tlsConfig itself and fields in tls.Config have not json tag for unmarshalling from JSON config.

**Comment by daheige** [comment]
The original implementation here is read-write timeout. The bottom layer of the redis component library is 5S. For a slow operation, it is almost tens of milliseconds. Here, refer to the parameter configuration of the bottom layer go redis.


------------------ Original ------------------
From: andrew shan ***@***.***&gt;
Date: Sun,Jun 5,2022 10:22 PM
To: polarismesh/polaris ***@***.***&gt;
Cc: heige ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [polarismesh/polaris] feat: refactor redispool (PR #408)





 
@andrewshan commented on this pull request.
 
 
In common/redispool/redis_pool.go:
 &gt;  }    // DefaultConfig redis pool configuration with default values  func DefaultConfig() *Config {  return &amp;Config{  MaxIdle:        200, -IdleTimeout:    commontime.Duration(120 * time.Second), -ConnectTimeout: commontime.Duration(300 * time.Millisecond),  
默认连接时长和消息超时为什么要改成5s？

**Comment by chuntaojun** [comment]
> There is no problem with the design here. You can understand it by looking carefully at the design of with options
> […](#)
> ------------------ Original ------------------ From: liaochuntao ***@***.***&gt; Date: Sun,Jun 5,2022 10:37 PM To: polarismesh/polaris ***@***.***&gt; Cc: heige ***@***.***&gt;, Author ***@***.***&gt; Subject: Re: [polarismesh/polaris] feat: refactor redispool (PR #408) @chuntaojun commented on this pull request. In common/redispool/config_option.go: &gt; +package redispool + +import ( +"crypto/tls" +"time" + +commontime "github.com/polarismesh/polaris-server/common/time" +) + +// Option functional options for Config +type Option func(c *Config) + +// WithConfig set new config for NewPool,keep old code compatibility +func WithConfig(newConfig *Config) Option { +return func(c *Config) { +c = newConfig 这个逻辑，应该是问题的？并不能真正影响c的实际内容吧 — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you authored the thread.Message ID: ***@***.***&gt;

简单写了个例子测试了下，测试的结果并不能按照这个方法实际期望的，

```golang
type testOption func(t *TestObject)

type TestObject struct {
	Method string
}

func NewTestObject(opts ...testOption) *TestObject {
	t := new(TestObject)

	for i := range opts {
		opts[i](t)
	}

	return t
}

func WithTestObjectMethod(m string) testOption {
	return func(t *TestObject) {
		t.Method = m
	}
}

func WithTestObject(newT *TestObject) testOption {
	return func(t *TestObject) {
		t = newT
	}
}

func TestWithNewAuthContext(t *testing.T) {
	ctxOne := NewTestObject(WithTestObjectMethod("method-1"))
	ctxTwo := NewTestObject(WithTestObjectMethod("method-2"))
	ctxOne = NewTestObject(WithTestObject(ctxTwo))
	assert.Equal(t, ctxOne, ctxTwo)
}
```

File: common/redispool/config_option.go (hunk 2 above)
Comment (chuntaojun) [reply common/redispool/config_option.go:16]: 你的写成 *c = *newConfig 才可以吧

File: common/redispool/redis_pool.go (hunk 1 above)
Comment (andrewshan) [reply common/redispool/redis_pool.go:114]: 默认值和心跳场景有关系，都恢复成300ms吧

**Comment by daheige** [comment]
Ok&nbsp;


------------------ Original ------------------
From: andrew shan ***@***.***&gt;
Date: Mon,Jun 6,2022 11:18 AM
To: polarismesh/polaris ***@***.***&gt;
Cc: heige ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [polarismesh/polaris] feat: refactor redispool (PR #408)





 
@andrewshan commented on this pull request.
 
 
In common/redispool/redis_pool.go:
 &gt;  }    // DefaultConfig redis pool configuration with default values  func DefaultConfig() *Config {  return &amp;Config{  MaxIdle:        200, -IdleTimeout:    commontime.Duration(120 * time.Second), -ConnectTimeout: commontime.Duration(300 * time.Millisecond),  
默认值和心跳场景有关系，都恢复成300ms吧

**Comment by shichaoyuan** [comment]
whether there is a need to add certificate configuration?

**Review by chuntaojun — APPROVED** [review]

**Comment by andrewshan** [comment]
@daheige hello, pls help to resolve the confilcts

**Pull request merged**